libsolidity/analysis/ImmutableValidator.h

```diff
@@ -0,0 +1,88 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libsolidity/ast/ASTVisitor.h>
+#include <liblangutil/ErrorReporter.h>
+
+#include <memory>
+
+namespace solidity::frontend
+{
+
+// Traverses the constructors and state variable assignments to builds a
+// function call graph
+class ImmutableValidator: private ASTConstVisitor
+{
+	struct CompareDef
+	{
+		using is_transparent = void;
+
+		bool operator()(Declaration const* _lhs, Declaration const* _rhs) const
+		{
+			return _lhs->id() < _rhs->id();
```

Comment:
Can we please move that to ASTNode as the default comparison operator?

Reply:
Ah, the problem is that it is a pointer, right? Then at least add a struct `CompareByID` to ASTNode or something.